File: shared/node_libs/src/crypto/tokens.ts
import { randomBytes } from 'crypto';

/**
 * Generates a cryptographically secure random token of the specified byte length.
 * Returns hex-encoded string.
 */
export function generateToken(length: number): string {
  if (length <= 0) {
    throw new Error(`Token length must be positive, got ${length}`);
  }

  return randomBytes(length).toString('hex');
}

/**
 * Generates a cryptographically secure random token of the specified byte length.
 * Returns URL-safe base64 encoded string (without padding).
 */
export function generateUrlSafeToken(length: number): string {
  if (length <= 0) {
    throw new Error(`Token length must be positive, got ${length}`);
  }

  return randomBytes(length)
    .toString('base64')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=/g, '');
}

File: shared/go_libs/crypto/tokens.go
package crypto

import (
	"crypto/rand"
	"encoding/base64"
	"encoding/hex"
	"fmt"
)

// GenerateToken generates a cryptographically secure random token of the specified byte length.
// Returns hex-encoded string.
func GenerateToken(length int) (string, error) {
	if length <= 0 {
		return "", fmt.Errorf("token length must be positive, got %d", length)
	}

	bytes := make([]byte, length)
	if _, err := rand.Read(bytes); err != nil {
		return "", fmt.Errorf("failed to generate random token: %w", err)
	}

	return hex.EncodeToString(bytes), nil
}

// GenerateURLSafeToken generates a cryptographically secure random token of the specified byte length.
// Returns URL-safe base64 encoded string (without padding).
func GenerateURLSafeToken(length int) (string, error) {
	if length <= 0 {
		return "", fmt.Errorf("token length must be positive, got %d", length)
	}

	bytes := make([]byte, length)
	if _, err := rand.Read(bytes); err != nil {
		return "", fmt.Errorf("failed to generate random token: %w", err)
	}

	return base64.RawURLEncoding.EncodeToString(bytes), nil
}
